### PR TITLE
Adding support to `DiracDelta` for generic points in the domain 

### PR DESCRIPTION
**Pull request by kishore-nori**
Hi @santiagobadia and @amartinhuertas, I added methods to `DiracDelta` to support generic points in the domain, which are not grid entities with tags, would be glad to know your suggestions for improvement and changes. This PR is aimed at resolving the issue [#357]. Given that `CellField`s can now be evaluated at a generic point in the domain, the method make use of this to calculate the integral of the composition of `DiracDelta` with a `CellField`, and also works for generic functions compatible to evaluate at a `Gridap` `Point`. 

So as to conform with `DiracDelta` struct, a `DiscreteModel` is constructed with given set of `Point`s to generate `Triangulation` and `Measure`. If we are not required to conform with the existing framework, we could just store the `Point`s as a struct and evaluate the `CellField` point-wise directly, but the above doesn't create much overhead. And probably the later is not generic enough to be extendable to line and surface forcing terms. 

The usage is same as earlier but the constructor is as follows:

```julia
model = CartesianDiscreteModel((0,1,0,1),(3,3))
p = Point(0.1,0.2)
pvec = [p, 1.5*p]
δ = DiracDelta(model,p)
δ_comb = DiracDelta(model,pvec)
```

[WIP] TO DO: support for line and surface loads! (like presented here https://github.com/gridap/Gridap.jl/issues/408#issuecomment-734929575). I think the line load case is extendable in a straight forward manner using `CartesianDiscreteModel` (but may be needs to be done separately for each line - defined by a pair of points..)

### CODE REVIEW

**Comment by amartinhuertas** [comment]
HI @kishore-nori ! Thanks for your work. Can you please add an entry to the `NEWS.md` file pointing to this work ?

File: test/CellDataTests/DiracDeltasTests.jl [hunk 1]
```diff
@@ -37,6 +37,22 @@ degree = 2
 δ = DiracDelta{0}(model,tags=[2,4])
 @test sum(δ(v)) ≈ 5
 
+# Tests for DiracDelta at a generic Point in the domain #
+
+p = VectorValue(0.2,0.3)
```
Comment (amartinhuertas) [inline test/CellDataTests/DiracDeltasTests.jl:42]: why do you use `VectorValue` and not `Point`?

File: test/CellDataTests/DiracDeltasTests.jl [hunk 2]
```diff
@@ -37,6 +37,22 @@ degree = 2
...
+δ = DiracDelta(p,model)
+@test sum(δ(v)) == v(p)
```
Comment (amartinhuertas) [inline test/CellDataTests/DiracDeltasTests.jl:44]: Replace `==` by `\sim` ?

File: test/CellDataTests/DiracDeltasTests.jl [hunk 3]
```diff
@@ -37,6 +37,22 @@ degree = 2
...
+
+pvec = [p,π*p]
+δ = DiracDelta(pvec,model)
+@test sum(δ(v)) == sum(v(pvec))
```
Comment (amartinhuertas) [inline test/CellDataTests/DiracDeltasTests.jl:48]: see https://github.com/gridap/Gridap.jl/pull/837/files#r994035461

File: test/TensorValuesTests/IndexingTests.jl [hunk 4]
```diff
@@ -14,6 +14,7 @@ v = VectorValue{4}(a)
 @test size(v) == (4,)
 @test length(v) == 4
 @test lastindex(v) == length(v)
+@test v[end] == a[end]
```
Comment (amartinhuertas) [inline test/TensorValuesTests/IndexingTests.jl:17]: where do these tests come from? shouldnt be these already in `master` ? (i.e., is your branch up to date with `master`?) they seem to be unrelated to this PR.

File: test/CellDataTests/DiracDeltasTests.jl (hunk 1 above)
Comment (kishore-nori) [reply test/CellDataTests/DiracDeltasTests.jl:42]: `Point` is not available for usage here, is has been defined in `Gridap.Fields` so just used `VectorValue` (as it seemed to me that the tests are also following the module hierarchy).

File: test/CellDataTests/DiracDeltasTests.jl [hunk 5]
```diff
@@ -37,6 +37,22 @@ degree = 2
...
+
+# below passes but cannot use FESpace and FEFunction here
```
Comment (amartinhuertas) [inline test/CellDataTests/DiracDeltasTests.jl:50]: Not sure why this commented code is here. Should it be here or we can just remove it?

File: test/TensorValuesTests/IndexingTests.jl (hunk 4 above)
Comment (kishore-nori) [reply test/TensorValuesTests/IndexingTests.jl:17]: No, I forgot adding these tests, sorry took the opportunity to add it now (instead of again creating a PR), sorry for the confusion, it is not under PR.

File: src/CellData/DiracDeltas.jl [hunk 6]
```diff
@@ -76,3 +79,33 @@ end
...
+  cache1 = _point_to_cell_cache(KDTreeSearch(),trian)
+  cell = map(x -> _point_to_cell!(cache1, x), v)
+  point_grid = UnstructuredGrid(v)
+  point_model = UnstructuredDiscreteModel(point_grid)
+  point_trian = Triangulation(point_model)
+  dx = Measure(point_trian,1)
+  DiracDelta{0,NotGridEntity}(point_trian,dx)
+end
+
+function evaluate!(cache,d::DiracDelta{0,NotGridEntity},f::CellField)
```
Comment (amartinhuertas) [inline src/CellData/DiracDeltas.jl:107]: why do we need this specialized version of `evaluate!`. Doesn't the generic one work straight away?

File: test/CellDataTests/DiracDeltasTests.jl [hunk 7]
```diff
@@ -37,6 +37,22 @@ degree = 2
...
+# reffe = ReferenceFE(lagrangian,Float64,3)
+# V = FESpace(model,reffe,conformity=:L2)
+# uh = FEFunction(V,rand(num_free_dofs(V)))
+# @test sum(d(uh)) == uh(p)
+
 #using Gridap
```
Comment (amartinhuertas) [inline test/CellDataTests/DiracDeltasTests.jl:56]: should we also test DiracDeltas with `FEFunction` and `FEBasis` ?

File: test/CellDataTests/DiracDeltasTests.jl (hunk 5 above)
Comment (kishore-nori) [reply test/CellDataTests/DiracDeltasTests.jl:50]: This is a good test but as I mentioned in the comment, I cant use `FESpace` and `FEFunction` here, as they are down the hierarchy. If it is ok I can use `using Gridap` for the test

**Review by amartinhuertas — CHANGES_REQUESTED** [review]
Hi @kishore-nori ! I have finished a first round of review. Please address or rebut my comments/remarks/etc..

File: test/CellDataTests/DiracDeltasTests.jl (hunk 7 above)
Comment (kishore-nori) [reply test/CellDataTests/DiracDeltasTests.jl:56]: Yes, liked I replied to one of the above comments, we need to include the corresponding modules to use the `FEFunction` and `FEBasis`, I would be happy to include these tests..

File: src/CellData/DiracDeltas.jl (hunk 6 above)
Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl:107]: No.. it results in quadrature and `FEFunction/CellField` not having the same/compatible background triangulation error

File: test/CellDataTests/DiracDeltasTests.jl (hunk 2 above)
Comment (kishore-nori) [reply test/CellDataTests/DiracDeltasTests.jl:44]: Yes I can do that, should be safer..

File: test/CellDataTests/DiracDeltasTests.jl (hunk 3 above)
Comment (kishore-nori) [reply test/CellDataTests/DiracDeltasTests.jl:48]: (replied above)

File: test/CellDataTests/DiracDeltasTests.jl (hunk 5 above)
Comment (amartinhuertas) [reply test/CellDataTests/DiracDeltasTests.jl:50]: >  I cant use FESpace and FEFunction here, as they are down the hierarchy. 

I dont understand this comment. Can't you just add `using Gridap.FESpaces` to the module in that test file?

Comment (kishore-nori) [reply test/CellDataTests/DiracDeltasTests.jl:50]: Yes I can.., my question was does `Gridap` `tests` also follow the module usage hierarchy as the `src`? If not I can do it.

File: test/CellDataTests/DiracDeltasTests.jl (hunk 1 above)
Comment (amartinhuertas) [reply test/CellDataTests/DiracDeltasTests.jl:42]: I dont see any problem in `using Gridap.Fields` at the test. (perhaps I am missing something)

**Comment by kishore-nori** [comment]
Thank you very much for the comments Alberto, I ll make changes accordingly following our discussion in comments here.

File: test/CellDataTests/DiracDeltasTests.jl (hunk 5 above)
Comment (amartinhuertas) [reply test/CellDataTests/DiracDeltasTests.jl:50]: I dont think we have to constrain ourselves to the same module usage hierarchy. In Gridap this has to be respected because otherwise we may end with dependcy loops (cycles) among modules.

You can think as the test directory as an external program that is using Gridap tools and features. Thus, I do not see why we are forced to respect that constraint.

File: src/CellData/DiracDeltas.jl (hunk 6 above)
Comment (amartinhuertas) [reply src/CellData/DiracDeltas.jl:107]: Ok. I now understand. The integrand is defined on a domain which is not (currently) compatible with the domain of the measure. But this may mean that there is a compatibility rule among triangulations that is somehow missing, because why these domains should not be compatible?

Comment (amartinhuertas) [reply src/CellData/DiracDeltas.jl:107]: This is just a though ... it you have the impression that this overly complicates thing, we may stick into the current workaround.

File: test/TensorValuesTests/IndexingTests.jl (hunk 4 above)
Comment (amartinhuertas) [reply test/TensorValuesTests/IndexingTests.jl:17]: ok, no worries ... we may merge them here, but the cleaner way to go is to do it in a separate PR

File: test/CellDataTests/DiracDeltasTests.jl (hunk 5 above)
Comment (amartinhuertas) [reply test/CellDataTests/DiracDeltasTests.jl:50]: The other thing that we have to think is whether `DiracDeltas.jl` should still be within `CellData` now given that we have introduced additional dependencies.

File: test/CellDataTests/DiracDeltasTests.jl (hunk 1 above)
Comment (kishore-nori) [reply test/CellDataTests/DiracDeltasTests.jl:42]: No problem will use it now, I was in a misconception that even the tests follow the same module include hierarchy as the Gridap source code does.

File: test/TensorValuesTests/IndexingTests.jl (hunk 4 above)
Comment (kishore-nori) [reply test/TensorValuesTests/IndexingTests.jl:17]: Yes I am sorry, will be careful next time and separate it into different PR. Probably would be too messy here now, to revert back?

File: src/CellData/DiracDeltas.jl (hunk 6 above)
Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl:107]: Initially I tried to make it work with the same triangulations but it wasn't generalising properly due to `CellQuadrature` constructing extra-cells of `#undef` values in `DomainContribution`. 

The problem occurs at @check is_change_possible(trian_f,trian_x) msg in the integrate method (https://github.com/gridap/Gridap.jl/blob/2a4921b473fe113b9d4b6168b8728a7a708f7bc6/src/CellData/CellQuadratures.jl#L142)

Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl:107]: > But this may mean that there is a compatibility rule among triangulations that is somehow missing, because why these domains should not be compatible?
Yes I was wondering about this as well.. will have to dig deeper

File: test/CellDataTests/DiracDeltasTests.jl (hunk 5 above)
Comment (kishore-nori) [reply test/CellDataTests/DiracDeltasTests.jl:50]: > The other thing that we have to think is whether DiracDeltas.jl should still be within CellData now given that we have introduced additional dependencies.
No.. I haven't introduced any dependency in DiracDeltas.jl, used only the existing dependencies

Comment (amartinhuertas) [reply test/CellDataTests/DiracDeltasTests.jl:50]: > No.. I haven't introduced any new dependency in DiracDeltas.jl, used only the existing dependencies

ok, I think you are right. Having a (new) reference to `CellField` (in the new `evaluate` function) does not imply a new module dependency, correct?

File: test/TensorValuesTests/IndexingTests.jl (hunk 4 above)
Comment (amartinhuertas) [reply test/TensorValuesTests/IndexingTests.jl:17]: Better not to revert back, I agree. 

For completeness, perhaps we can add in the NEWS.md a comment that we added some missing tests missing from another PR.

File: test/CellDataTests/DiracDeltasTests.jl (hunk 5 above)
Comment (kishore-nori) [reply test/CellDataTests/DiracDeltasTests.jl:50]: Yes.. since `CellField` belongs to `CellData` module itself, it is not a new or external dependency.

File: test/TensorValuesTests/IndexingTests.jl (hunk 4 above)
Comment (kishore-nori) [reply test/TensorValuesTests/IndexingTests.jl:17]: Definitely, will mention it in the `NEWS.md`

**Comment by kishore-nori** [comment]
Hi @amartinhuertas and @santiagobadia, I have rewritten the implementation (realised I could do it in a good way, after our discussion to approach it as quadrature). This now works for evaluations and also solving the PDE with `DiracDelta` at a single or even a set of `Points`. I added tests to compare with both the exact solution and the earlier tag based implementation of `DiracDelta`. I also incorporated the earlier review comments from our discussions. Would be happy to know further comments for improvement.

File: src/CellData/DiracDeltas.jl [hunk 8]
```diff
@@ -65,14 +70,63 @@ function DiracDelta{0}(model::DiscreteModel;tags)
...
+  cell_points = collect(values(cell_to_pindices))
+  points = map(i->pvec[cell_points[i]], 1:length(cell_ids))
+  weights_x_cell = collect.(Fill.(one(T),length.(cell_points)))
+  pquad = map(i -> GenericQuadrature(pvec[cell_points[i]],weights_x_cell[i]), 1:length(cell_ids))
+  trianv = Triangulation(trian,cell_ids)
+  pmeas = Measure(CellQuadrature(pquad,points,weights_x_cell,trianv,PhysicalDomain(),PhysicalDomain()))
+  GenericDiracDelta{0,D,NotGridEntity}(trianv,pmeas)
+end
+
+function evaluate!(cache,d::GenericDiracDelta{0,Dt,NotGridEntity},f::Function) where Dt
```
Comment (amartinhuertas) [inline src/CellData/DiracDeltas.jl:120]: Which is the ultimate reason behind having to define this specialised method? 

Why cant we leverage the generic implementation of `evaluate!` whenever the domain of `DiracDelta` is a function?

File: src/CellData/DiracDeltas.jl [hunk 9]
```diff
@@ -65,14 +70,63 @@ function DiracDelta{0}(model::DiscreteModel;tags)
...
+  cell_to_pindex
+end
+
+function DiracDelta(model::DiscreteModel{D}, p::Point{D,T}) where {D,T}
+  trian = Triangulation(model)
+  cache = _point_to_cell_cache(KDTreeSearch(),trian)
+  cell = _point_to_cell!(cache, p)
+  trianv = TriangulationView(trian,[cell])
+  pquad = GenericQuadrature([p],[one(T)])
+  pmeas = Measure(CellQuadrature([pquad],[[p]],[[one(T)]],trianv,PhysicalDomain(),PhysicalDomain()))
```
Comment (amartinhuertas) [inline src/CellData/DiracDeltas.jl:103]: to be on the safe side w.r.t. memory allocations, better not to create `[p]` and `[one(T)]` twice.

why do we have to pass `[[p]]` to `CellQuadrature` constructor? Couldn't it get this info from `pquad` ? (perhaps a missing constructor in Gridap?)

**Review by amartinhuertas — CHANGES_REQUESTED** [review]
Hi @kishore-nori ! nice work! I finished with my review. Please answer/rebut my comments.

File: src/CellData/DiracDeltas.jl (hunk 8 above)
Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl:120]: I was in a misconception that existing `evaluate!` method goes through the search process of Point to Cell (which ofcourse works perfectly fine as well), which is only necessary for `CellFields`, Functions can be evaluated directly without the search. But it seems it doesn't go through that route once we have the `CellPoint` available, which is very nice. So I ll remove it..

File: src/CellData/DiracDeltas.jl (hunk 9 above)
Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl:103]: Yes true, I was thinking the same, may be a missing constructor, specifically for `data_domain_style` being `PhysicalDomain()`, the problem is we can't use the `compress_cell_data` for `AbstractArrays` only for `Fill` and `CompressedArray`, which is the usual quadrature case. Yes I ll remove the repetition and reuse.

File: src/CellData/DiracDeltas.jl (hunk 8 above)
Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl:120]: But one advantage of the new constructor is it is ~4 times faster and aloocates ~10 times less memory.

Comment (amartinhuertas) [reply src/CellData/DiracDeltas.jl:120]: > But it seems it doesn't go through that route once we have the CellPoint available, which is very nice.

Whenever you have a `CellPoint`, Gridap assumes that the points in each cell are within the cell (either reference cell or physical cells, depending on the value of the corresponding trait). 

If the points are not within the cell, then I guess you have an undefined behaviour. But we are satisfying this precondition, as the cells have been judiciously selected in the TriangulationView such that the hold the corresponding points.

Comment (amartinhuertas) [reply src/CellData/DiracDeltas.jl:120]: > But one advantage of the new constructor is it is ~4 times faster and aloocates ~10 times less memory.

Do these ratios keep asympotatically fixed as you refine the mesh?

Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl:120]: > Whenever you have a CellPoint, Gridap assumes that the points in each cell are within the cell (either reference cell or physical cells, depending on the value of the corresponding trait).

>If the points are not within the cell, then I guess you have an undefined behaviour. But we are satisfying this precondition, as the cells have been judiciously selected in the TriangulationView such that they hold the corresponding points.

Yes true.. since we are checking it, it should be safe, it is just a matter of efficiency

Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl:120]: > Do these ratios keep asympotatically fixed as you refine the mesh?
I ll check and update.

Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl:120]: On refinement the the new version runs in the same time but the earlier one becomes slower and more memory intensive.

File: src/CellData/DiracDeltas.jl (hunk 9 above)
Comment (amartinhuertas) [reply src/CellData/DiracDeltas.jl]: > Yes true, I was thinking the same, may be a missing constructor, specifically for data_domain_style being PhysicalDomain(), the problem is we can't use the compress_cell_data for AbstractArrays only for Fill and CompressedArray, which is the usual quadrature case. Yes I ll remove the repetition and reuse.

Ok, I see. The constructor is missing, but if we added it, we would not be able to leverage it in our particular context (we do not have Fill or CompressedArrays to represent the quadrature points and weights).

File: src/CellData/DiracDeltas.jl [hunk 10]
```diff
@@ -65,14 +70,65 @@ function DiracDelta{0}(model::DiscreteModel;tags)
...
+    push!(get!(() -> valtype(cell_to_pindex)[], cell_to_pindex, cell), i)
+  end
+  cell_to_pindex
+end
+
+function DiracDelta(model::DiscreteModel{D}, p::Point{D,T}) where {D,T}
+  trian = Triangulation(model)
+  cache = _point_to_cell_cache(KDTreeSearch(),trian)
+  cell = _point_to_cell!(cache, p)
+  trianv = TriangulationView(trian,[cell])
```
Comment (amartinhuertas) [inline src/CellData/DiracDeltas.jl:101]: use `view` Julia method instead of `TriangulationView`. 

This way we do not need to export `TriangulationView`.

File: src/Geometry/Geometry.jl [hunk 11]
```diff
@@ -102,6 +102,7 @@ export compress_ids
 export UnstructuredGridTopology
 
 export Triangulation
+export TriangulationView
```
Comment (amartinhuertas) [inline src/Geometry/Geometry.jl:105]: see https://github.com/gridap/Gridap.jl/pull/837/files#r996625877

File: src/CellData/DiracDeltas.jl [hunk 12]
```diff
@@ -65,14 +70,65 @@ function DiracDelta{0}(model::DiscreteModel;tags)
...
+
+function DiracDelta(model::DiscreteModel{D}, pvec::Vector{Point{D,T}}) where {D,T}
+  trian = Triangulation(model)
+  cell_to_pindices = _cell_to_pindices(pvec,trian)
+  cell_ids = collect(keys(cell_to_pindices))
+  cell_points = collect(values(cell_to_pindices))
+  points = map(i->pvec[cell_points[i]], 1:length(cell_ids))
+  weights_x_cell = collect.(Fill.(one(T),length.(cell_points)))
+  pquad = map(i -> GenericQuadrature(points[i],weights_x_cell[i]), 1:length(cell_ids))
+  trianv = TriangulationView(trian,cell_ids)
```
Comment (amartinhuertas) [inline src/CellData/DiracDeltas.jl:117]: https://github.com/gridap/Gridap.jl/pull/837/files#r996625877

File: src/CellData/DiracDeltas.jl [hunk 13]
```diff
@@ -65,14 +70,65 @@ function DiracDelta{0}(model::DiscreteModel;tags)
...
+  pmeas = Measure(CellQuadrature(pquad,points,weights_x_cell,trianv,PhysicalDomain(),PhysicalDomain()))
+  GenericDiracDelta{0,D,NotGridEntity}(trianv,pmeas)
+end
+
+function evaluate!(cache,d::GenericDiracDelta{0,Dt,NotGridEntity},f::Function) where Dt
```
Comment (amartinhuertas) [inline src/CellData/DiracDeltas.jl:122]: Remove this method.

File: src/Geometry/Geometry.jl (hunk 11 above)
Comment (kishore-nori) [reply src/Geometry/Geometry.jl]: done!

File: src/CellData/DiracDeltas.jl (hunk 12 above)
Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl]: done!

File: src/CellData/DiracDeltas.jl (hunk 13 above)
Comment (kishore-nori) [reply src/CellData/DiracDeltas.jl]: done :)
